toolchain/lex/tokenized_buffer.cpp

```diff
@@ -49,20 +55,35 @@ auto VariantMatch(V&& v, Fs&&... fs) -> decltype(auto) {
   return std::visit(Overload{std::forward<Fs&&>(fs)...}, std::forward<V&&>(v));
 }
 
-#if __x86_64__
-#define CARBON_USE_SIMD 1
-
+#if CARBON_USE_SIMD
 // A table of masks to include 0-16 bytes of an SSE register.
-static const std::array<__m128i, sizeof(__m128i) + 1> prefix_masks = [] {
-  std::array<__m128i, sizeof(__m128i) + 1> masks = {};
-  for (auto [i, mask] : llvm::enumerate(masks)) {
-    memset(&mask, 0xFF, i);
-  }
-  return masks;
-}();
+static const
+#if __ARM_NEON
+    std::array<uint8x16_t, sizeof(uint8x16_t) + 1>
+#elif __x86_64__
+    std::array<__m128i, sizeof(__m128i) + 1>
 #else
-#define CARBON_USE_SIMD 0
+#error "Unsupported SIMD architecture!"
+#endif
+        prefix_masks = [] {
+          using MaskArrayT = std::remove_const_t<decltype(prefix_masks)>;
+          using MaskT = MaskArrayT::value_type;
+          MaskArrayT masks = {};
+          for (auto [i, mask] : llvm::enumerate(masks)) {
+            std::array<uint8_t, sizeof(MaskT)> mask_bytes = {};
+            memset(&mask_bytes, 0xFF, i);
+#if __ARM_NEON
+            mask = vld1q_u8(mask_bytes.data());
+#elif __x86_64__
+            mask = _mm_loadu_si128(
+                reinterpret_cast<const __m128i*>(mask_bytes.data()));
```

Comment:
Given `MaskT` on line 73, would it help readability if this used the same even though it's in a platform `#if`?

Reply:
I generally like having the platform specific types when using the platform specific intrinsics as it makes how things are working more clear for me.

But this is also somewhat obviated -- the suggestion to make this `constexpr` not just works, but works portably and so this is now portable code using the typedef. =]